### ISSUE TITLE
Is it really a compat package?

### ISSUE THREAD

**Issue by narfbg** [paragonie/random_compat#7]
If this is indeed a compatibility package, it should behave exactly like PHP does (when possible of course). Therefore, it should check for (and give the highest priority to) /dev/arandom availability too.

Yes, /dev/urandom will simply redirect to /dev/arandom on current systems that have it and [I made the same point on internals](https://marc.info/?l=php-internals&m=142477156804904), but I also got [a somewhat reasonable response](https://marc.info/?l=php-internals&m=142477544706473) on that.

On a side note, while I do agree that exceptions are better in this case, I'm quite sure that PHP itself currently returns FALSE (and emits an E_WARNING) in case that `random_bytes()` fails. You might want to work with them to change that while there's still time. :) Also ... I'd use InvalidArgumentException and RuntimeException instead of the base Exception class.

**Comment by paragonie-scott** [comment]
Thanks for your careful review :)

> Therefore, it should check for (and give the highest priority to) /dev/arandom availability too.

Fair enough, I'm currently working on a commit for some of the recommendations in #6 and will add that while I'm at it.

> On a side note, while I do agree that exceptions are better in this case, I'm quite sure that PHP itself currently returns FALSE (and emits an E_WARNING) in case that `random_bytes()` fails.

See these two PRs:
* https://github.com/php/php-src/pull/1397
* https://github.com/php/php-src/pull/1398

If either or both of them is rejected, I will amend this library to reflect this design decision.

**Comment by narfbg** [comment]
Great. :)

**Comment by paragonie-scott** [comment]
Unfortunately, there is one compatibility issue I don't think we can safely resolve: https://github.com/paragonie/random_compat/commit/d27f2e67faf252b12c21299ed4f685ba16ecaa7c

**Comment by oittaa** [comment]
How about using [BC Math](https://secure.php.net/manual/en/book.bc.php) to calculate the range? It complicates the function a bit, because you have to convert those bit shift operations to ```bcmul``` and ```bcdiv```, but it should work.

**Comment by narfbg** [comment]
@paragonie-scott Yea, as I told you on twitter, I wanted to do a similar package myself ... couldn't come-up with a proper solution for `random_int()`.

But BC Math may help, I guess.

**Comment by oittaa** [comment]
Adding support for whole integer range wasn't that hard. ```random_int(-PHP_INT_MAX - 1, PHP_INT_MAX);```

Here's output from diff -u
```
@@ -149,26 +150,27 @@
         if ($min > $max) {
              throw new Exception('Minimum value must be less than or equal to the maximum value');
         }
-        $range = $max - $min;
+        $range = bcsub($max, $min);
 
         /**
          * Test for integer overflow:
          */
-        if (!is_int($range)) {
-             throw new Exception('An integer overflow occurred on the range between $min and $max');
-        }
+//        if (!is_int($range)) {
+//             throw new Exception('An integer overflow occurred on the range between $min and $max');
+//        }
 
         /**
          * Do we have a meaningful range? If not, return the minimum value.
          */
-        if ($range < 1) {
+        if (bccomp($range, '1') === -1) {
             return $min;
         }
 
         /**
          * Initialize variables to 0
          */
-        $rejections = $bits = $bytes = $mask = 0;
+        $rejections = $bits = $bytes = 0;
+        $mask = '0';
 
         $tmp = $range;
         /**
@@ -180,13 +182,13 @@
          * $bits is effectively ceil(log($range, 2)) without dealing with 
          * type juggling
          */
-        while ($tmp > 0) {
+        while (bccomp($tmp, '0') === 1) {
             if ($bits % 8 === 0) {
                ++$bytes;
             }
             ++$bits;
-            $tmp >>= 1;
-            $mask = $mask << 1 | 1;
+            $tmp = bcdiv($tmp, '2');
+            $mask = bcadd(bcmul($mask, '2'), '1');
         }
 
         /**
@@ -210,27 +212,27 @@
              * Let's turn $rval (random bytes) into an integer
              * 
              * This uses bitwise operators (<< and |) to build an integer
-             * out of the values extracted from ord()
+             * out of the values extracted from ord() and applies the mask
              * 
              * Example: [9F] | [6D] | [32] | [0C] =>
              *   159 + 27904 + 3276800 + 201326592 =>
              *   204631455
              */
-            $val = 0;
+            $val = '0';
+            $mask_tmp = $mask;
             for ($i = 0; $i < $bytes; ++$i) {
-                $val |= (ord($rval[$i]) << ($i * 8));
+                $tmp = (int)bcmod($mask_tmp, '256');
+                $mask_tmp = bcdiv($mask_tmp, '256');
+                $val = bcadd($val, bcmul((string)(ord($rval[$i]) & $tmp), bcpow('2', (string)($i*8))));
             }
 
-            // Apply mask
-            $val &= $mask;
-
-            if ($val > $range) {
+            if (bccomp($val, $range) === 1) {
                 ++$rejections;
             }
             // If $val is larger than the maximum acceptable number for
             // $min and $max, we discard and try again.
-        } while ($val > $range);
-        return (int) ($min + $val);
+        } while (bccomp($val, $range) === 1);
+        return (int) bcadd($min, $val);
     }
 }
 ```

**Comment by oittaa** [comment]
IMHO this
```
             if (bccomp($val, $range) === 1) {
                 ++$rejections;
             }
```
could be replaced with just ```++$rejections;```. We shouldn't really care that the counter gets incremented anyway.

**Comment by paragonie-scott** [comment]
Do any of the BC math functions open any timing side-channels that we should be aware of?

**Comment by CodesInChaos** [comment]
@paragonie-scott I believe that my implementation handles the whole range, without using big integer arithmetic.

**Comment by paragonie-scott** [comment]
@CodesInChaos I tried implementing your `$reject` algorithm, but I think I made a mistake which led to #6

**Comment by paragonie-scott** [comment]
@CodesInChaos Aha, I figured out what I did wrong last time. I just took another swing and got `random_int(-PHP_INT_MAX, PHP_INT_MAX)` to work :).

**Comment by paragonie-scott** [comment]
https://github.com/paragonie/random_compat/issues/7#issuecomment-119650294
Changed in
https://github.com/paragonie/random_compat/commit/84e9832c71665849edf20df111b4634ed9d8bb9e

**Comment by paragonie-scott** [comment]
https://wiki.php.net/rfc/random-function-exceptions

Given that this is so far unanimous, it seems highly likely that this will pass. As soon as the PRs are updated/merged I'll close these issues, run through final checks, then tag 1.0.0.

https://github.com/paragonie/random_compat/commit/15571f580bbdf6d594d18a97c9582ba19d23bbfa

**Comment by geggleto** [comment]
This is excellent news. Thank you for the update :+1:

**Comment by paragonie-scott** [comment]
Well, this *is* really a compat library now.

So far, the votes are 27 to 1 in favor and, most relevantly, the pull requests against php-src have been updated. Now I just need to update the contributors list on the README and wait until Sunday morning and this can be closed and v1.0.0 can be tagged.

**Comment by narfbg** [comment]
Great. :)

**Comment by paragonie-scott** [comment]
https://wiki.php.net/rfc/random-function-exceptions 28-2, https://github.com/php/php-src/pull/1397 should be merged soon, so random_compat 1.0.0 incoming tonight.
